### ISSUE TITLE
glitching presets

### ISSUE THREAD

**Issue by severak** [mikerodd/june-21#8]
I was playing with June-21 and I noticed some of the presets are glitching or not working at all.

This got better after last update - nothing crash synth anymore. 

However there some sounds which I believe has some glitches in it:

- P15 - Lead Synth - no usable sound - just glitchy sizzle
- P18 - ElectroDrm - hi freq clinking sometimes
- P51 - Brass - sometimes hi freq hiss
- P67 - SynthBass2 - usable but lots of annoying sizzle
- P67 - SynthKoto - sometimes hi freq hiss
- M15 - Trumpets - hi freq hiss in background
- M16 - BrassSwell - hi freq hiss sometimes
- M17 - PolySynth2 - hi freq hiss sometimes
- M18 - PolySynth3 - hi freq hiss only on bass notes
- M21 - BowdStrngs - hi freq hiss sometimes
- M48 - Harmonica - hi freq hiss always
- M54 - ChrusPluck - hi freq cling sometimes
- M55 - Bells2 - hi freq cling sometimes
- M62 - Lead 6 - sizzle and pops when changing notes
- M66 - SynthBass3 - easily overdriven - then pops a lot
- M78 - OctaveJump - not sure if doing what it is intended - sounds more like a motor than anything musical

This hi freq hiss (mentioned many times) is IMHO from filter. It's more noticable when playing full chords on bass notes.

Also note that I have no physical Juno to compare to, so this is just my observations driven fully by my ears.

**Comment by mikerodd** [comment]
Thanks a lot for that, It's surely have to do with the filters, I'm trying to get rid of it

**Comment by mikerodd** [comment]
P15 is good now (problem in reverse dco envelope)
hi freq hiss should be better now, I suppressed some correcting filter
M78  : still an issue related to enveloppe (need to work on this)
M62 : still a problem too :-/

**Comment by eisbaer23** [comment]
About the high frequency ringing noise...
I was thinking a small DC offset from the DCOs, but even when you turn all of them off, it's still there.
Except when you use the HPF on settings 2 and 3. Which again, seems to point out a DC offset, mmm.
I modified the HPF 1 (aka no HPFiltering at all) to "mvchpf aOutVcoblock,1" and it's gone.
So it would seem some noise from the VCO Block (wait, isn't that supposed to be a DCO, I'm confused now) is leaking thru anyway.

**Comment by eisbaer23** [comment]
There's definitely some leaking going on from the sub osc.
Open June-21, leave it to P-11. Turn OFF the Pulse and Sawtooth. No sound.
Now dial up the Sub Level to 1, play a note, turn it back to 0. Eww...
Turn it up to 3, play a note, back to 0 again. Ewwwwww... ;-)

**Comment by eisbaer23** [comment]
Added a "else aOsc3 = 0" at around line 818 and that did the trick for me.
Is my thinking wrong?
To me it looks like there is/was no proper cleaning of aOsc3 in the whole "if (kSubLevel != 0)"

**Comment by mikerodd** [comment]
Yes, you're right there is a clear improvement (no click nor hissing !) :-) 
Great! 
:-)

**Comment by mikerodd** [comment]
I still have problems in some sound due to enveloppe emulation problem (M78 - OctaveJump)

**Comment by eisbaer23** [comment]
Well, there's still some clicking and hissing when an oscillator is on, so clearly it's not THE fix. But there was some subosc stuff going on that made it worse, even when unused!!!
I still think there is some audio leakage or build up from the oscillator block but despite some research I haven't been able to figure it out. There's math in it. Ew... ;-)

**Comment by eisbaer23** [comment]
There are clicks and then harmonics during the ramping (T1, T2 & T4) phases of the envelope, What I don't understand is, they seem to be fixed (1500 Hz and octaves below/above), as in not note-dependent, yet they seem to follow the DCO range. They are filtered by both the HPF and LPF, but always present. They disappear at the sustain L3 phase, tho. Any idea where they're from?

**Comment by eisbaer23** [comment]
Found it! It's ksmps.
Using "ksmps = 1", no more hiss and less clicks. ;-) It does tax the CPU a wee bit more tho.
Not familiar with CSound but it seems like there are audio-rate and k-rate types of envelopes and maybe audio-rate is smoother. Unless there is a way to smooth the envelope as they are?

**Comment by mikerodd** [comment]
Yes a-rate are supposed to be smoother with the `interop` opcode
I'll give a try with various values of ksmps (but I have almost no hisses at home ?)
=> could you send me a sound example of the problem to see (listen in fact ^^) in order to try to reproduce at home ? 

Thanks !

**Comment by eisbaer23** [comment]
Maybe those will do. Had some problems getting the older 0.9.3 to work, I guess some internal parameters were changed regarding the use of presets as it went tits up. Also I forgot to fix the 48kHz in the 0.9.3 render, so, bad tuning. Oops.
I think the DCblock2 might not be that helpful afterall, it changes the sound too much and it seems the interpolation on the ENV is enough. Tho I didn't make tests with and without yet.
[harmonics.zip](https://github.com/mikerodd/june-21/files/4111490/harmonics.zip)
So anyway, the interpolation does a great job at reducing those zipper noise/harmonics during the ramping and it's almost perfect at this point. Yes, maybe upping the ksmps might help a bit more, but it's almost not worth the cpu, or maybe just when rendering? Maybe there is a way to change this automagically when rendering versus live mode in a DAW...
![093_p88_timps](https://user-images.githubusercontent.com/59757208/73114398-58383680-3f1a-11ea-9e2e-54bd02e35ac2.png)
